src/charm.py

```diff
@@ -103,6 +106,30 @@ def __init__(self, *args):
         self.framework.observe(self.on.web_pebble_ready, self._configure)
         self.framework.observe(self.on.pfcp_agent_pebble_ready, self._configure)
         self.framework.observe(self.on.config_changed, self._configure)
+        self.framework.observe(
+            self.fiveg_n3_provider.on.fiveg_n3_request, self._update_fiveg_n3_relation_data
+        )
+
+    def _update_fiveg_n3_relation_data(self, event: EventBase) -> None:
+        """Publishes UPF IP address in the `fiveg_n3` relation data bag.
+
+        Args:
+            event: Juju event
+        """
+        if not self._access_ip_config_is_valid():
+            self.unit.status = BlockedStatus("Invalid `access-ip` config provided")
+            return
+        upf_access_ip_address = self._get_access_network_ip_config().split("/")[0]  # type: ignore[union-attr]  # noqa: E501
+        fiveg_n3_relations = self.model.relations.get("fiveg_n3")
+        if not fiveg_n3_relations:
+            logger.info("No `fiveg_n3` relations found.")
+            return
+        for fiveg_n3_relation in fiveg_n3_relations:
+            if self._get_upf_ip_from_relation_data_bag(fiveg_n3_relation) != upf_access_ip_address:
```

Comment:
should we add a test case to validate this? (when `self._get_upf_ip_from_relation_data_bag(fiveg_n3_relation) == upf_access_ip_address`